### ISSUE TITLE
Voxel Cubes not texturing. and meshes laggy and unpreforment.

### ISSUE THREAD

**Issue by Derpysphere** [Zylann/godot_voxel#662]
No matter what I try my voxel cubes will not texture. I have a 1024 by 1024 texture sheet, I have 5 types of blocks:

Air
Grass
Dirt
Sand
Sand Alt

If I attempt to give them a texture, (by giving the material a texture.) the're just white ingame. I'm not sure whether I'm just texturing them wrong or I setup the engine wrong. or perhaps something else?

But I'm curious what the proper method for texturing them is.

I looked into the voxel demo is listed, but there you seem to be using the meshes, instead of cubes, I tried this for myself but Godot died of graphical overwork. As it seems the Cube meshes render faces that aren't even visible to the camera. or perhaps not. once again, I'm not familiar enough with the engine to know. although I have the setting for rendering hidden faces turned off something is going wrong.

**Comment by Zylann** [comment]
It's all explained here: https://voxel-tools.readthedocs.io/en/latest/blocky_terrain/

> the're just white ingame

My first suspicion without more context is that you have set a material override on the terrain itself, which will override materials from the library. Usually you don't want a material directly on the terrain.

> As it seems the Cube meshes render faces that aren't even visible to the camera

You probably made them wrong. There are strict rules to make meshes as shown in the docs. If those rules aren't followed, one of the outcomes is that no sides get culled.

If you still can't find what's wrong in your project, you could try sending a minimal test project showing your issue.

**Comment by Derpysphere** [comment]
Awesome, thanks.

**Comment by Zylann** [comment]
Is your problem fixed?

**Comment by Derpysphere** [comment]
Yep. The mesh lag had to do with the world gen. and the texture problem was exactly what you thought. originally when testing I set the world generation to have a override material, then promptly forgot about it. Thanks again.
